coffea/nanoevents/mapping.py

```diff
@@ -119,6 +176,212 @@ def __getitem__(self, key):
+                out = awkward1.layout.NumpyArray(out)
+            else:
+                raise Exception('array is not flat array or jagged list')
+            return awkward1.Array(out)
+
+    def __init__(self, fileopener, cache=None, access_log=None):
+        super(ParquetSourceMapping, self).__init__(fileopener, cache, access_log)
+
+    @classmethod
+    def _extract_base_form(cls, arrow_schema):
+        column_forms = {}
+        for field in arrow_schema:
+            key = field.name
+            fmeta = field.metadata
+
+            if "," in key or "!" in key:
+                warnings.warn(
+                    f"Skipping {key} because it contains characters that NanoEvents cannot accept [,!]"
+                )
+                continue
+
+            form = None
+            if b'form' in fmeta:
+                form = json.loads(fmeta[b'form'])
+            else:
+                schema = field.type
+                form = arrow_schema_to_awkward_form(schema)
+                form = json.loads(form.tojson())
+
+            if (
+                form["class"].startswith("ListOffset")
+                and form["content"]["class"] == "NumpyArray"  # noqa
+            ):
+                form["form_key"] = quote(f"{key},!load")
+                form["content"]["form_key"] = quote(f"{key},!load,!content")
+                if b'title' in fmeta:
+                    form["content"]["parameters"] = {"__doc__": fmeta[b'title'].decode()}
+                elif "__doc__" not in form["content"]["parameters"]:
+                    form["content"]["parameters"] = {"__doc__": key}
+            elif form["class"] == "NumpyArray":
+                form["form_key"] = quote(f"{key},!load")
+                if b'title' in fmeta:
+                    form["parameters"] = {"__doc__": fmeta[b'title'].decode()}
+                elif "__doc__" not in form["parameters"]:
+                    form["parameters"] = {"__doc__": key}
+            else:
+                warnings.warn(
+                    f"Skipping {key} as it is not interpretable by NanoEvents"
+                )
+                continue
+            column_forms[key] = form
+        return {
+            "class": "RecordArray",
+            "contents": column_forms,
+            "parameters": {"__doc__": "parquetfile"},
+            "form_key": "",
+        }
+
+    def key_root(self):
+        return "ParquetSourceMapping:"
+
+    def preload_column_source(self, uuid, path_in_source, source):
+        """To save a double-open when using NanoEventsFactory.from_file"""
+        key = self.key_root() + tuple_to_key((uuid, path_in_source))
+        self._cache[key] = source
+
+    def get_column_handle(self, columnsource, name):
+        return ParquetSourceMapping.UprootLikeShim(columnsource, name)
+
+    def extract_column(self, columnhandle, start, stop):
+        # make sure uproot is single-core since our calling context might not be
```

Comment:
stray comment

Reply:
Reminds me I need to turn the parquet multi-core read off by default!